### PR TITLE
configuration option to disable watch commands

### PR DESCRIPTION
**Pull request by terabyte**
Suggested in #544.  This is my first stab at this feature.  Looking for
feedback about what to do when commands are disabled (print ERROR,
other?).  Examples of other feature flags I should look at?

TODO:
* test locally
* write tests
* get feedback on help option name / descriptions
* other?

### CODE REVIEW

**Comment by dormando** [comment]
The flush_all command has a similar disable flag, so you can model it the same way.

**Comment by terabyte** [comment]
So I have a pass above including tests even, but unfortunately the tests hang.  I implemented it looking at the way flush_all is disabled (so it uses auth failure in the case where a watch command is attempted).  The current tests as written hang for 600s then time out.  I think this is because the auth failure does not include a newline at the end to flush the buffer?  I'm kinda stuck on that right now:
```
$  telnet localhost 11211
Trying ::1...
Connected to localhost.
Escape character is '^]'.
watch
Auth failure.
```
Note there is no newline after the "Auth failure." line.  I think this is why my tests are hanging?  Here is what the output currently looks like:
```
$ prove -v t/watcher.t
t/watcher.t .. 
1..13
ok 1 - watcher enabled
ok 2 - basic get works
ok 3 - skipped some lines
ok 4 - saw a real log line after a skip
ok 5 - stored the long key
ok 6 - got the correct uri encoded key length
ok 7 - new watcher enabled
ok 8 - evictions watcher enabled
ok 9 - found rawcmd log entry
ok 10 - found eviction log entry
ok 11 - mutations watcher enabled
ok 12 - correctly logged cas command
Timeout.. killing the process
Timeout.. killing the process
not ok 13 - attempted watch gave auth failure with no_watch option
#   Failed test 'attempted watch gave auth failure with no_watch option'
#   at t/watcher.t line 133.
Auth failure.': ' 
#     expected: 'Auth failure.'
# Looks like you failed 1 test of 13.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/13 subtests 

Test Summary Report
-------------------
t/watcher.t (Wstat: 256 Tests: 13 Failed: 1)
  Failed test:  13
  Non-zero exit status: 1
Files=1, Tests=13, 603 wallclock secs ( 0.02 usr  0.01 sys +  0.16 cusr  0.44 csys =  0.63 CPU)
Result: FAIL
```

**Comment by terabyte** [comment]
aaaaaaand I figured it out.  There are two protocols, the binary one and the text one.  I copied the implementation for the binary protocol that disables flush_all and used that in the text protocol code for watch.  This has been rectified now.

I didn't see any watch code for the binary protocol, does it not support watch commands?  if the answer to that is "yes", then I think this change is good to go!

File: memcached.c [hunk 1]
```diff
@@ -5577,6 +5578,12 @@ static void process_watch_command(conn *c, token_t *tokens, const size_t ntokens
     assert(c != NULL);
 
     set_noreply_maybe(c, tokens, ntokens);
+    if (settings.no_watch) {
+        // watch is not allowed
```
Comment (dormando) [inline memcached.c]: redundant comment.

File: memcached.h [hunk 2]
```diff
@@ -428,6 +428,7 @@ struct settings {
     unsigned int logger_watcher_buf_size; /* size of logger's per-watcher buffer */
     unsigned int logger_buf_size; /* size of per-thread logger buffer */
     bool drop_privileges;   /* Whether or not to drop unnecessary process privileges */
+    bool no_watch; /* allows watch commands to be dropped */
```
Comment (dormando) [inline memcached.h]: why "no_watch" vs "watch_enabled", like flush_enabled?

**Comment by dormando** [comment]
I never bothered implementing watch for binprot. I'm deprecating the entire protocol in favor of the new meta commands anyway.

couple nits, looks fine otherwise. thanks!

File: memcached.h (hunk 2 above)
Comment (terabyte) [reply memcached.h]: I think I saw a "no_foo" flag elsewhere.  Happy to reverse this if you prefer.  I also noticed that the flush feature was a regular option, not an extended option, but I had already started to write this as an extended option - would you like me to change that or is an extended option fine?

File: memcached.c (hunk 1 above)
Comment (terabyte) [reply memcached.c]: will remove.

File: memcached.h (hunk 2 above)
Comment (dormando) [reply memcached.h]: hmm I guess the options are kind of inconsistent. Lets make it a top level option if there's an obvious free character? Most of the extendeds are tuning or very rarely used.

**Comment by dormando** [comment]
@terabyte I disappeared for longer than expected. still interested in finishing this PR?

**Comment by dormando** [comment]
I flipped this around to the -W option, didn't change anything else. Merged into `next` now

**Comment by dormando** [comment]
out in 1.5.21, thanks!

**Pull request closed without merging**